<b>Hola Hans</b>

Mi nombre es Alejandro Abia y tengo el gusto de revisar tu proyecto.

A continuación, encontrarás mis comentarios en celdas pintadas de tres colores (verde, amarillo y rojo), a manera de semáforo. Por favor, <b>no las borres ni muevas de posición</b> mientras dure el proceso de revisión.

<div class="alert alert-block alert-success">
<b>Éxito</b> <a class="tocSkip"></a>
En celdas verdes encontrarás comentarios en relación a tus aciertos y fortalezas.
</div>
<div class="alert alert-block alert-warning">
<b>Antención</b> <a class="tocSkip"></a>
Utilizaré el color amarillo para llamar tu atención, expresar algo importante o compartirte alguna idea de valor.
</div>
<div class="alert alert-block alert-danger">
<b>A resolver</b> <a class="tocSkip"></a>
En rojo emitiré aquellos puntos que deberás atender para aprobar la revisión.
</div>
<div class="alert alert-block alert-info">
<b>Comentario estudiante</b><a class="tocSkip"></a>
Es factible que, a lo largo del proceso de revisión, quieras dejarme comentarios. Si es el caso, por favor realízalo dentro de celdas azules como esta.
</div>
Respecto del proceso de revisión, tu proyecto será aceptado una vez que los comentarios en rojo hayan sido atendidos.
¡Empecemos!

# Déjame escuchar música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa de ellos y tomar decisiones en consecuencia. Esto implica pasar por diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investiguemos, necesitamos formular hipótesis que después podamos probar. A veces, aceptaremos estas hipótesis; otras veces, las rechazaremos. Para tomar las decisiones adecuadas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de música online para probar la hipótesis que planteamos a continuación y compararás el comportamiento de los usuarios de estas dos ciudades.

### Objetivo:
Probar la hipótesis:
1. La actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento de los usuarios se almacenan en el archivo `/datasets/music_project_en.csv`. No hay información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, debes evaluar la calidad de los datos y ver si los problemas son significativos. Más tarde, durante el preprocesamiento de datos, deberás abordar los problemas más críticos.

Tu proyecto contará con estas tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de los datos.
 3. Prueba de la hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [2]:

import pandas as pd

Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [3]:
df = pd.read_csv('/datasets/music_project_en.csv')


Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [4]:
print(df.head(10))



     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [5]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-success">
<b>Celdas [3–5]</b> <a class="tocSkip"></a><br>
Muy buena ejecución de la carga y exploración inicial del conjunto de datos. Usar <code>df.head()</code> e <code>df.info()</code> permitió comprender rápidamente la estructura de los datos, su tipo y la presencia de valores nulos. Estas prácticas son fundamentales en cualquier análisis.
</div>


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1.todos los datos son objet, al final de la tabla info indica que tipo de dato almacena 

2.hay 65,079 filas, pero con valores nulos, si hay suficiente informacion pero se puede ver afectada por los valores nulos

3.primero hay que organizar la informacion para su tratamiento y posterior hay que tratar los datos nulos para saber que tan confiables son o como afectan el resultado 

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
nombres_columnas = df.columns.tolist()



In [24]:
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case.


Anteriormente, aprendiste una forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora.

Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [25]:
# Bucle en los encabezados que lo pone todo en minúsculas
new_columns = []
for column in df.columns:
    new_columns.append(column.lower())


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [29]:
# Bucle en los encabezados que elimina los espacios
new_columns = []
for column in df.columns:
    new_columns.append(column.strip().lower())
df.columns = new_columns

Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [31]:
# Cambia el nombre de la columna "userid"

df = df.rename(columns={'userid': 'user_id'})

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [32]:
# Comprueba el resultado: lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Celdas [25–32]</b> <a class="tocSkip"></a><br>
Excelente trabajo estandarizando los nombres de las columnas. La transición a snake_case, la eliminación de espacios y el uso coherente de minúsculas mejoran la limpieza y mantenibilidad del dataset.
</div>


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [33]:
# Calcula el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan de la misma forma a la investigación. Por ejemplo, los valores ausentes en 'track' y 'artist' no son cruciales para el análisis, ya que estos datos son más descriptivos que analíticos. Por eso, puedes reemplazarlos directamente con un valor predeterminado como el string 'unknown' (desconocido).

En cambio, los valores ausentes en 'genre' sí pueden influir en la comparación entre las preferencias musicales de Springfield y Shelbyville. En un escenario real, lo ideal sería investigar por qué faltan estos datos e intentar recuperarlos. Pero en este proyecto no tenemos esa posibilidad, así que deberás rellenar esos valores con un valor predeterminado.


Como vimos anteriormente en las lecciones, la mejor manera de hacerlo es crear una lista con los nombres de las columnas que necesitan reemplazarse.  Luego, utilizar esta lista para iterar sobre cada columna y realizar el reemplazo correspondiente.

Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [34]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [35]:
# Cuenta los valores ausentes
print(df[columns_to_replace].isna().sum())

track     0
artist    0
genre     0
dtype: int64


<div class="alert alert-block alert-success">
<b>Celdas [34–35]</b> <a class="tocSkip"></a><br>
Muy bien resuelto el manejo de valores nulos. Usar <code>fillna('unknown')</code> para columnas descriptivas es una práctica adecuada cuando no se puede recuperar la información faltante.
</div>


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [36]:
# Cuenta los duplicados explícitos
duplicados_totales = df.duplicated().sum()

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [37]:
# Elimina los duplicados explícitos
df.drop_duplicates(inplace=True)

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [40]:
# Comprueba de nuevo si hay duplicados
duplicados_restantes = df.duplicated().sum()


print(duplicados_restantes)

0


<div class="alert alert-block alert-success">
<b>Celdas [36–40]</b> <a class="tocSkip"></a><br>
Buen uso de <code>duplicated()</code> y <code>drop_duplicates()</code> para manejar duplicados explícitos. El doble chequeo antes y después es una excelente práctica para validar que el proceso fue efectivo.
</div>


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [44]:
# Inspecciona los nombres de géneros únicos
unique_genres = df['genre'].unique()


sorted_genres = sorted(unique_genres)
for genre in sorted_genres:
    print(genre)

acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'heavy
hardcore
hardstyle
hardtechno
hip
hip-hop
hiphop
historisch
holiday
hop
horror
house
idm
i

Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Algunos de los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_genres()` que tendrá dos parámetros:

* `wrong_genres`: una lista con todos los valores que deben ser reemplazados.  
* `correct_genre`: un string que se utilizará como valor de reemplazo.

El objetivo de esta función es **corregir los valores en la columna `'genre'` del DataFrame `df`**, reemplazando cada valor de la lista `wrong_genres` por `correct_genre`.

Etapa 2.12.
1. Define una función llamada `replace_wrong_genres()` que reciba dos parámetros: `wrong_genres` y `correct_genre`.

2. Dentro de la función, utiliza un bucle `for` para iterar sobre cada valor en la lista `wrong_genres`.

3. En cada iteración, accede a la columna `'genre'` del DataFrame `df` y utiliza el método `.replace()` para sustituir el valor incorrecto por `correct_genre`.

4. Llama a la función y pasa como argumentos:
   - Una lista con los duplicados implícitos: `['hip', 'hop', 'hip-hop']`
   - El string de reemplazo: `'hiphop'`

In [45]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
     for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

Etapa 2.13. Ahora, llama a `replace_wrong_genres()` y pásale estos argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los sustituya por `hiphop`:

In [46]:
# Elimina los duplicados implícitos
replace_wrong_genres(
    wrong_genres=['hip', 'hop', 'hip-hop'], 
    correct_genre='hiphop')

<div class="alert alert-block alert-success">
<b>Celdas [45–46]</b> <a class="tocSkip"></a><br>
La creación de la función <code>replace_wrong_genres()</code> muestra comprensión de cómo aplicar transformaciones reutilizables sobre el DataFrame. También se nota atención al detalle en la limpieza de valores inconsistentes, lo cual es fundamental en análisis reales.
</div>


Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [50]:
# Comprueba de nuevo los duplicados implícitos
print(sorted(df['genre'].dropna().unique()))


['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

### Observaciones <a id='data_preprocessing_conclusions'></a>

`Redacta un breve resumen de lo que descubriste al analizar los datos. Tu respuesta debe identificar los problemas detectados, explicar cómo los resolviste y describir cómo esas acciones mejoran la calidad del análisis.`

Etapa 2.14. Descríbelo aquí. bueno a lo largo del curso hemoa visto diferentes listas de datos y sabemos que en su mayoria son recabadas con informacion que el mismo cliente brinda y dado a que es un proceso automatico hay veces que la informacion no esta completa o es incorrecta y en el caso de esta lista de datos podemos ver errores de agrupacion, de datos faltantes y errores en el nombre de las columnas lo ideal es que los datos ya vayan limpios para poder ahorrar tiempo y enfocarnos mas en el analisis de los datos 

[Volver a Contenidos](#back)

## Etapa 3. Prueba de la hipótesis <a id='hypothesis'></a>

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

La hipótesis que queremos probar plantea que existen diferencias en la forma en que los usuarios de Springfield y Shelbyville consumen música.
Para analizar esto, nos enfocaremos en los datos correspondientes a tres días específicos de la semana: lunes, miércoles y viernes.

El análisis consistirá en comparar la cantidad de canciones reproducidas por los usuarios de cada ciudad en esos días. Esto nos permitirá observar posibles patrones o diferencias en los hábitos de consumo entre Springfield y Shelbyville.

Para llevar a cabo este análisis, es importante seguir el enfoque de dividir-aplicar-combinar, del que ya hablamos en la lección. En este caso:

*  Dividir: separamos los datos en grupos según la ciudad.

*  Aplicar: dentro de cada grupo, contamos cuántas canciones se reprodujeron.

*  Combinar: reunimos los resultados en una estructura que nos permita comparar fácilmente ambas ciudades.

Este procedimiento debe repetirse de forma independiente para cada uno de los tres días seleccionados. El resultado final debe mostrar el número de reproducciones por ciudad en cada uno de esos días.

Una posible forma de estructurar el código sería con la siguiente expresión:

`df.groupby(by='...')['...'].method()`

Deberás completar los argumentos correspondientes para agrupar por ciudad y contar las canciones reproducidas. Este enfoque te dará una visión clara y comparativa del comportamiento de los usuarios en ambas ciudades.

Etapa 3.1. Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna 'track' como referencia.

In [54]:
# Cuenta las canciones reproducidas en cada ciudad
plays_by_city = df.groupby('city')['track'].count()
print(plays_by_city)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`Redacta brevemente tus observaciones sobre los resultados. ¿Qué diferencias notaste entre Springfield y Shelbyville? ¿A qué podrían deberse esas diferencias?`

Etapa 3.2. Escribe tus observaciones aquí. pues veo una gran diferencia y dado que no conozco las ciudades y probalemente sea imaginarias (simpsons) solo puedo decir que la diferencia puede ser por tama;o de poblacion

Etapa 3.3.
1. Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes, miércoles y viernes.

2. Utiliza el mismo método de conteo que antes, pero ahora cambia la columna de agrupación para enfocarte en el día.

3. Esto te permitirá identificar posibles patrones de escucha según el día de la semana.


In [57]:
# Calcula las canciones reproducidas en cada uno de los tres días
target_days = ['Monday', 'Wednesday', 'Friday']

for day in target_days:
    day_data = df[df['day'] == day]
    plays_by_city = day_data.groupby('city')['track'].count()
    print(day)
    print(plays_by_city)

Monday
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64
Wednesday
city
Shelbyville     7003
Springfield    11056
Name: track, dtype: int64
Friday
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64


<div class="alert alert-block alert-success">
<b>Celdas [54–57]</b> <a class="tocSkip"></a><br>
Has aplicado correctamente el enfoque dividir-aplicar-combinar para agrupar por ciudad y por día. El análisis por día permite observar patrones de comportamiento, y los resultados están bien organizados.
</div>


`Describe brevemente qué observaste al comparar los lunes, miércoles y viernes. ¿Hubo algún día con menos actividad? ¿Cambian las conclusiones si analizas cada ciudad por separado?`

Etapa 3.4. Escribe tus observaciones aquí.pues despues de una vista rapida de los datos y confirmando que en Springfiel tienen mayor numero de personas que escuchan musica solo puedo decir que el dia con mas personas es miercoles para Shelbyville , mientras que para Springfield es el viernes 

Hasta ahora has aprendido a contar entradas agrupándolas por un solo criterio, como la ciudad o el día de la semana. Ahora vamos a dar un paso más: necesitas crear una función que **cuente cuántas canciones se reprodujeron en una ciudad específica durante un día determinado**, combinando ambos criterios de filtrado.

La función se llamará `number_tracks()` y debe aceptar dos parámetros:

- `day`: un día de la semana (por ejemplo, `'Monday'`).
- `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`).

Dentro de la función, deberás aplicar un **filtrado secuencial mediante indexación lógica**: primero tendrás que filtrar el DataFrame por el día, y luego por la ciudad. Una vez que tengas el subconjunto de datos correcto, debes contar cuántas veces aparece un valor en la columna `'user_id'`. Ese número representará el total de canciones reproducidas bajo esos dos criterios.



Etapa 3.5.
1. Declara una función llamada `number_tracks()` con dos parámetros: `day` y `city`.

2. Filtra el DataFrame para conservar solo las filas donde la columna `'day'` sea igual al valor del parámetro `day`.

3. A partir del resultado anterior, filtra nuevamente para conservar solo las filas donde la columna `'city'` sea igual al valor del parámetro `city`.

4. Extrae la columna `'user_id'` del DataFrame filtrado y utiliza el método `.count()` para contar el número de entradas.

5. Guarda ese valor en una variable y **devuélvelo** como resultado de la función.


In [59]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):

    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    day_rows = df[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    city_day_rows = day_rows[day_rows['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    play_count = city_day_rows['user_id'].count()
    # Devuelve el número de valores de la columna 'user_id'
    return play_count

Etapa 3.6. Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que puedas recuperar los datos de ambas ciudades para cada uno de los tres días.

In [60]:
# El número de canciones reproducidas en Springfield el lunes
print("Springfield:", number_tracks('Monday', 'Springfield'))

Springfield: 15740


In [61]:
# El número de canciones reproducidas en Shelbyville el lunes
print("Shelbyville:", number_tracks('Monday', 'Shelbyville'))

Shelbyville: 5614


In [62]:
# El número de canciones reproducidas en Springfield el miércoles
print("Springfield:", number_tracks('Wednesday', 'Springfield'))

Springfield: 11056


In [63]:
# El número de canciones reproducidas en Shelbyville el miércoles
print("Shelbyville:", number_tracks('Wednesday', 'Shelbyville'))

Shelbyville: 7003


In [64]:
# El número de canciones reproducidas en Springfield el viernes
print("Springfield:", number_tracks('Friday', 'Springfield'))

Springfield: 15945


In [65]:
# El número de canciones reproducidas en Shelbyville el viernes
print("Shelbyville:", number_tracks('Friday', 'Shelbyville'))

Shelbyville: 5895


<div class="alert alert-block alert-success">
<b>Celdas [59–65]</b> <a class="tocSkip"></a><br>
Muy buen diseño y uso de la función <code>number_tracks()</code>. Muestra dominio de la indexación lógica y el uso de filtros múltiples. La repetición sistemática para distintos casos demuestra un enfoque organizado.
</div>


[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

##Escribe tus conclusiones finales sobre la hipótesis.
 * ¿Los datos apoyan la idea de que el comportamiento de los usuarios respecto a la música que escuchan varía según la ciudad y el día de la semana?
 * Indica si la hipótesis debe aceptarse o rechazarse, y justifica tu respuesta con base en los resultados obtenidos.

Etapa 4. Detalla aquí tus conclusiones. Springfield tuvo ciertamente más reproducciones que Shelbyville en los tres días analizados, creo que estos datos solo nos muestarn una diferencia sea geografica o cultural tendriamos que analizar otros factores para tomar decisiones mas importantes 
La hipótesis se acepta. Los datos demuestran que

Factores geográficos (ciudad) influyen en el consumo musical, posiblemente debido a diferencias demográficas, preferencias culturales u oferta de eventos locales.


### Nota
En los proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)

<div class="alert alert-block alert-success">
<b>Comentario final</b> <a class="tocSkip"></a><br>
¡Gran trabajo, Hans! El proyecto demuestra una muy buena comprensión del flujo de análisis de datos: desde la carga y limpieza hasta la prueba de hipótesis. Tus funciones están bien estructuradas, el código es claro y se aplican buenas prácticas en cada etapa. ¡Sigue así!
</div>
